Derived stores composition + tiny optimizations

pushkine opened the pull request:
I made some tweaks to svelte stores in a standalone for another project and thought it might be worth a PR

The current `derived` store function performs unnecessary work by merging 4 runtimes into a single one. It can either have one or multiple stores and `set` is either called automatically through a returned value or manually where the returned value becomes an optional cleanup. I rewrote it to compose from 2 pairs of functions instead. 

It gives derived stores a clear structure that makes the logic behind it easily understandable. While I did not change any logic or introduce any new behavior, I do find the modularity of that design to be very inspiring.

Also did tiny changes in `writable`, removed some variables & added some types

antony commented on src/runtime/store/index.ts:
```diff
@@ -66,46 +66,37 @@ export function writable<T>(value: T, start: StartStopNotifier<T> = noop): Writa
-			}
+		if (!safe_not_equal(value, new_value)) return;
+		value = new_value;
+		if (!stop) return;
```
As much as I'm absolutely 100% against comments, it's not my call here, and we've lost a comment. I think it should be either reinstated, or better:

```js
const store_is_ready = stop
```

In a perfect world:

```js
const store_is_ready = stop
if (!store_is_ready) { return }
```

However I realise this is runtime code, so I'm guessing there is an emphasis on bytes down the wire + terseness over readability. I don't know how TS optimises all of this stuff, so therefore maybe the comment is the perfect solution here. Ideally though, I'd aim for readability of the code.

pushkine commented:
added explicit returns
removed const creation in loop
replaced `=== 0` with `!` and `!== -1` with `~`

everything else is opinion and won't impact runtime, it's just a c/c from a project of mine
maintainers can edit so feel free to rearrange things to your liking

pushkine commented:
### derived
 * [x] rewrote for least amount of runtime work
 * [x] split into comprehensible categories
 * [x] rewrote some typescript ( should really swap ts for jsdoc tbh, this is _awful_ )
 * [ ] get typescript to fill types properly for the default, the `auto` and the `manual` ( need a pro to take a quick look on this one, it barely needs tweaking I guess )

pushkine commented:
`'svelte/internal'` fails to import, I think it has to do with the type import I added in `utils.ts` for the `subscribe` function, I'm not sure what to change in the rollup file to solve that
